map/bookmark_manager.cpp

```diff
@@ -2976,6 +2977,132 @@ UserMarkLayer * BookmarkManager::GetGroup(kml::MarkGroupId groupId)
   return it != m_categories.end() ? it->second.get() : nullptr;
 }
 
+struct QuadtreeElement
+{
+  int m_rank;
+  Bookmark * m_bm;
+  uint64_t m_zCurveCoord = 0; // z-order curve coordiante
+
+  bool operator < (QuadtreeElement const & rhs) const
+  {
+    return m_zCurveCoord < rhs.m_zCurveCoord;
+  }
+
+  static bool rankLess(QuadtreeElement const & lhs, QuadtreeElement  const & rhs)
+  {
+    return lhs.m_rank < rhs.m_rank;
+  }
+};
+
+using Quadtree = std::vector<QuadtreeElement>;
+
+static void SetMinZoom(Quadtree & quadtree)
+{
+  if (quadtree.empty())
+    return;
+
+  m2::RectD bbox;
+  for (auto const & [rank, bm, zCurveCoord] : quadtree)
+  {
+    UNUSED_VALUE(rank);
+    UNUSED_VALUE(zCurveCoord);
+    bbox.Add(bm->GetPivot());
+  }
+
+  if (bbox.IsEmptyInterior())
+    return;
+
+  auto const size = std::max(bbox.SizeX(), bbox.SizeY());
+  static constexpr int kZoomDelta = 4;  // approx 16x16 bookmarks maximum per drawn tile
+  static constexpr int kMaxZoom = scales::GetUpperStyleScale();
+  int const baseZoomLevel = std::max(-std::ilogb(std::clamp(size / mercator::Bounds::kRangeX, 0.0, 1.0)) - kZoomDelta, 1);
+  if (baseZoomLevel >= kMaxZoom)
+    return;
+
+  // enlarge bounding box to minimal enclosing centered square
```

Comment:
Please rewrite all separate-line comments to start with a capital letter and end with a dot.